cache/cache.go

```diff
@@ -13,3 +13,135 @@
 // limitations under the License.
 
 package cache
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+)
+
+var (
+	ErrDisabled  = errors.New("cache disabled")
+	ErrCompacted = errors.New("revision compacted")
+)
+
+type Cache struct {
+	cfg Config
+
+	cli *clientv3.Client
+	es  *eventSource
+	dx  *demux
+
+	stop context.CancelFunc
+	wg   sync.WaitGroup
+
+	gapSem chan struct{}
+	closed atomic.Bool
+}
+
+func New(cli *clientv3.Client, opts ...Option) (*Cache, error) {
+	cfg := defaultConfig()
+	for _, o := range opts {
+		o(&cfg)
+	}
+	if cfg.Disable {
+		return &Cache{cfg: cfg}, nil
+	}
+
+	// discover current revision
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	resp, err := cli.Get(ctx, "\x00", clientv3.WithLimit(1))
+	cancel()
+	if err != nil {
+		return nil, err
+	}
+
+	parentCtx, stop := context.WithCancel(context.Background())
+
+	es := newEventSource(cli, resp.Header.Revision, parentCtx)
+
+	c := &Cache{
+		cfg:    cfg,
+		cli:    cli,
+		es:     es,
+		dx:     newDemux(),
+		stop:   stop,
+		gapSem: make(chan struct{}, cfg.MaxGapWatch),
+	}
+
+	c.wg.Add(1)
+	go func() {
+		defer c.wg.Done()
+		es.run(func(ev *clientv3.Event) { c.dx.broadcast(ev, cfg.DropThreshold) })
+	}()
+	return c, nil
+}
+
+func (c *Cache) Watch(ctx context.Context, prefix string, rev int64) (*Stream, error) {
```

Comment:
If you want to expose `Watch` method, it should have the same function prototype like client watch.